src/Symfony/Component/AccessToken/AccessToken.php

```diff
@@ -0,0 +1,83 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\AccessToken;
+
+/**
+ * @author Pierre Rineau <pierre.rineau@processus.org>
+ */
+class AccessToken implements AccessTokenInterface
+{
+    /**
+     * Identifier for tokens that are not generated using credentials.
+     */
+    public const IN_MEMORY = 'in_memory';
+
+    protected ?\DateTimeImmutable $expiresAt;
+    protected ?bool $hasExpired = null;
+
+    /**
+     * @param string $id Identifier of credentials used for generating it
+     */
+    public function __construct(
+        protected readonly string $value,
+        protected readonly string $type = 'Bearer',
+        protected readonly int $expiresIn = 600,
```

Comment:
Indeed, rightly spotted, I maybe have missed to document it's seconds!

The default value actually come from an history of successive copy-pastes, this is a value I found in more than one pieces of code, gists, and other code samples. It actually makes sense, it's 10 minutes, and I often in the paste stumbled upon 15 minutes lifetime tokens. When the provider doesn't expose a lifetime, we need an arbitrary value to set for the cache anyway.

If you have a better idea how to manage this, but ideally I'd like to keep a default here. It could be simply 0, to tell not to cache, but I prefer to be conservative here and cache it anyway when the provider erroneously didn't expose a lifetime, so that this token will be retried at least once.

Reply:
it could make sense to have an option (on the factory, not the token) to provide a default value. and as the token is always created by the factory (from how i understood) the token could chose to not provide any defaults.

and maybe even an option to force a value for working with bad implementation (or not, could also ask user to extend the factory for such edge cases).

Reply:
@dbu Making the default lifetime configurable is what I would want to do, but I have a very theorical question about this.

There is basically two options:

 - Either make it configurable at the provider level, but as it is now, providers are not configurable, they are dead simple. Adding some common configuration options on `AbstractProvider` would either mean force new parameter constructors (or setters) and make them pass all the way from the concrete implementation to the abstract implementation. I don't like this, I do it when it makes sense, but I don't think it does here.
 - Either make it configurable on a per-credentials basis, I would tend to allow user to give a different default lifetime depending upon the remote service, not the provider base implementation, it's much more flexible!

So this poses a new question, fetchers are stateless, more or less, access token manager is too, so the credentials class must carry it, here again, two options:

 - Either add a `CredentialsInterface::getDefaultLifetime(): int` and implement it over all credentials, then use it in the providers.
 - Either simply let each implementation does it by itself, i.e. using an additional GET parameter over the user-provider DSN (for example: `oauth://service.tld/oauth/token?client_id=foo&client_secret=bar&default_lifetime=600` and deal with that in each specialized credentials factory.

In all cases, I'd probably wish to keep the default value on the `AccessToken` class, but it would allow each credentials to carry its own value here.

I'd lean toward the second option here, what do you think ?

TL;DR:
 - Let the user add a default_lifetime=N option in its URL.
 - Let each specialized credentials factory deal with it.
 - Let each specialized provider deal with it.
 - No method added in interfaces.

Reply:
i agree that the DSN / symfony configuration is the place where the default should be configured.

i would have added an options array to the provider. using the OptionsResolver, you can validate and document them well, plus provide default values for them for when they are not provided. the default lifetime should be on the level that instantiates the access token.

ttl has little to do with credentials imho, i would not use the credentials to carry any additional information. i would add options to ClientCredentialsProvider and have that use the configuration instead of the hardcoded 600. (600 is a fine default for when no default is specified).

Reply:
ah, i see the credentials also provide the endpoint url. is the credentials ~ the value object for the DSN? then it would make sense to have the options on the credentials indeed.

Reply:
Yes, indeed, external services such as Microsoft will give you different token URL depending upon your tenant identifier, so one way or another, the user will have to write its URL in configuration. There's of course some cases for specific providers that would handle by themselves the service URL (which is planned using `foo_service://default?options=...`) but for a generic OAuth implementation, that's simply not possible.

Reply:
@dbu Sorry, it took me a while, I was absent for 10 days. I implemented it this way:

- Add `CredentialsInterface::getDefaultLifetime(): int` which is implemented per default on `AbstractCredentials` returning either an arbitrary constructor set property, or the default 600 value if null.
- This new method can be implemented by each provider to give a sensible value.
- Add `&default_lifetime=X` query parameter handling in existing factory in order to propagate the user given value, on a per-url basis.

Do you have any opinion or suggestion about this ?

Reply:
great, to me that seems like a clean and no-surprises solution that should cover all relevant scenarios.

Reply:
Great, resolve this conversation, feel free to reopen anytime.